src/contracts/GPv2Settlement.sol

```diff
@@ -100,24 +106,128 @@ contract GPv2Settlement {
+        returns (
+            GPv2AllowanceManager.Transfer[] memory inTransfers,
+            GPv2AllowanceManager.Transfer[] memory outTransfers
+        )
+    {
+        uint256 tradeCount = encodedTrades.tradeCount();
+        inTransfers = new GPv2AllowanceManager.Transfer[](tradeCount);
+        outTransfers = new GPv2AllowanceManager.Transfer[](tradeCount);
+
+        GPv2Encoding.Trade memory trade;
+        for (uint256 i = 0; i < tradeCount; i++) {
+            encodedTrades.tradeAtIndex(i).decodeTrade(
+                domainSeparator,
+                tokens,
+                trade
+            );
+            processTrade(
+                trade,
+                clearingPrices[trade.sellTokenIndex],
+                clearingPrices[trade.buyTokenIndex],
+                inTransfers[i],
+                outTransfers[i]
+            );
+        }
+    }
+
+    /// @dev Process trades for a single EOA order.
+    /// @param trade The trade to process.
+    /// @param sellPrice The price of the order's sell token.
+    /// @param buyPrice The price of the order's buy token.
+    /// @param inTransfer Memory location to set computed tranfer into the
+    /// settlement contract to execute trade.
+    /// @param outTransfer Memory location to set computed transfer out to order
+    /// owner to execute trade.
+    function processTrade(
+        GPv2Encoding.Trade memory trade,
+        uint256 sellPrice,
+        uint256 buyPrice,
+        GPv2AllowanceManager.Transfer memory inTransfer,
+        GPv2AllowanceManager.Transfer memory outTransfer
+    ) internal pure {
+        GPv2Encoding.Order memory order = trade.order;
+        // NOTE: Currently, the above instanciation allocates an unitialized
+        // `Order` that gets never used. Adjust the free memory pointer to free
+        // the unused memory by subtracting `sizeof(Order) == 288` bytes.
+        // <https://solidity.readthedocs.io/en/v0.7.5/internals/layout_in_memory.html>
+        // TODO: Remove this once the following fix is merged and released:
+        // <https://github.com/ethereum/solidity/pull/10341>
+        // solhint-disable-next-line no-inline-assembly
+        assembly {
+            mstore(0x40, sub(mload(0x40), 288))
+        }
+
+        inTransfer.owner = trade.owner;
+        inTransfer.token = order.sellToken;
+        outTransfer.owner = trade.owner;
+        outTransfer.token = order.buyToken;
+
+        if (order.kind == GPv2Encoding.OrderKind.Sell) {
+            uint256 executedSellAmount;
+            if (order.partiallyFillable) {
+                executedSellAmount = trade.executedAmount;
+            } else {
+                executedSellAmount = order.sellAmount;
+            }
+
+            inTransfer.amount = executedSellAmount;
+            // NOTE: Don't use `SafeMath.div` here as it allocates a string even
+            // if it does not revert.
+            outTransfer.amount = executedSellAmount.mul(buyPrice) / sellPrice;
```

Comment:
Is there value in trying to avoid overflows with largeish (>`uint128`, <`uint256`) amounts?
We can probably come up with something that eliminates unnecessary overflows at the price of
- one extra conditional statement per order (assuming there is no overflow for that order)
- additionally, a bunch of conditions, `%`, `*`, and `/` here and there if an overflow is expected).

Reply:
Division can't overflow, its just a `!= 0` check since dividing by 0 uses all remaining transaction gas while reverting just stops using gas at that point.

Reply:
Otherwise I'd be reluctant to prematurely optimize this without some sort of benchmarking or having a better idea of the kind of settlements we are expecting (what if the solver uses very large prices to be able to have more precise exchange rates?). But something we can look at when we are doing a once-over gas optimization.

Reply:
I'm ok with the `/`, the comment was about the `.mul`: if `executedSellAmount = 2**128` , `buyPrice = 2**128 `, and `sellPrice = 2**128 ` then the transaction reverts, even if the result of this operation fits a `uint256`. This could be avoided, but it's a bit of extra logic.